c/src/session.rs

```diff
@@ -21,20 +21,27 @@
 
 use std::ffi::c_char;
 
-use typedb_driver::{Database, Options, Session, SessionType};
+use typedb_driver::{DatabaseManager, Options, Session, SessionType};
+use crate::memory::string_view;
 
 use super::{
     error::{try_release, unwrap_void},
-    memory::{borrow, borrow_mut, free, release_string, take_ownership},
+    memory::{borrow, borrow_mut, free, release_string},
 };
 
 #[no_mangle]
 pub extern "C" fn session_new(
-    database: *mut Database,
+    databases: *mut DatabaseManager,
+    database_name: *const c_char,
     session_type: SessionType,
     options: *const Options,
 ) -> *mut Session {
-    try_release(Session::new_with_options(take_ownership(database), session_type, borrow(options).clone()))
+    borrow(databases).get(string_view(database_name)).map_or(
+        std::ptr::null_mut(),
+        |database| try_release(Session::new_with_options(
+            database,
+            session_type,
+            borrow(options).clone())))
```

Comment:
`try_release` already handles the error case for you, you don't need to provide `null_mut()` yourself. `try_release(databases.get(...).and_then(...))` should be enough.